stored_requests/backends/file_fetcher/fetcher.go

```diff
@@ -16,45 +16,93 @@ import (
 // This expects each file in the directory to be named "{config_id}.json".
 // For example, when asked to fetch the request with ID == "23", it will return the data from "directory/23.json".
 func NewFileFetcher(directory string) (stored_requests.Fetcher, error) {
-	storedReqData, err := collectStoredData(directory + "/stored_requests")
-	if err != nil {
-		return nil, err
-	}
-	storedImpData, err := collectStoredData(directory + "/stored_imps")
-	if err != nil {
-		return nil, err
-	}
-
-	return &eagerFetcher{storedReqData, storedImpData}, nil
+	storedData, err := collectStoredData(directory, FileSystem{make(map[string]FileSystem), make(map[string]json.RawMessage)}, nil)
+	return &eagerFetcher{storedData}, err
 }
 
 type eagerFetcher struct {
-	storedReqs map[string]json.RawMessage
-	storedImps map[string]json.RawMessage
+	FileSystem FileSystem
 }
 
 func (fetcher *eagerFetcher) FetchRequests(ctx context.Context, requestIDs []string, impIDs []string) (map[string]json.RawMessage, map[string]json.RawMessage, []error) {
-	errs := appendErrors("Request", requestIDs, fetcher.storedReqs, nil)
-	errs = appendErrors("Imp", impIDs, fetcher.storedImps, errs)
-	return fetcher.storedReqs, fetcher.storedImps, errs
+	storedRequests := fetcher.FileSystem.Directories["stored_requests"].Files
+	storedImpressions := fetcher.FileSystem.Directories["stored_imps"].Files
+	errs := appendErrors("Request", requestIDs, storedRequests, nil)
+	errs = appendErrors("Imp", impIDs, storedImpressions, errs)
+	return storedRequests, storedImpressions, errs
+}
+
+func (fetcher *eagerFetcher) FetchCategories(primaryAdServer, publisherId, iabCategory string) (string, error) {
+	if primaryAdServerDir, found := fetcher.FileSystem.Directories[primaryAdServer]; found {
+
+		fileName := primaryAdServer
+		if len(publisherId) != 0 {
+			fileName = primaryAdServer + "_" + publisherId
+		}
+
+		if file, ok := primaryAdServerDir.Files[fileName]; ok {
+
+			tmp := make(map[string]string)
+
+			if err := json.Unmarshal(file, &tmp); err != nil {
```

Comment:
Since this is called once per (winning) bid, this will cause us to unmarshal the same JSON multiple times per request. See above about unmarshalling at startup and storing that result.

Reply:
@hhhjort code refactored to unmarshall same categories file once. Please see previous comment